### PR TITLE
fix(eagle): restore dp=1 dense EAGLE3 path on main + add E2E to CI

### PR DESCRIPTION
**Pull request by zorrofox**
Part of #1053 (P1-0 prerequisite).

## Motivation

`speculative/` was last touched by #939 (4/28). Since then #939/#940/#1015 changed sampler / `schedule_batch` / `compilation_manager` that the EAGLE path depends on, with no EAGLE-side updates; `test_speculative_decoding.py` was not in `run_suite.py` so CI did not catch it. This restores the **dp=1 dense EAGLE3** path (Qwen3-32B + EAGLE3) end-to-end as the baseline for #1053 P1-2's class refactor.

### Regression fixes — #939 (`ScheduleBatch` flat → `reqs_info`)
- `ScheduleBatch.__getattr__/__setattr__` dp=1 compat shim passing through `reqs_info[0]` until #1053 P1-5 lands the DP-aware data contract
- `get_spec_model_worker_batch`: populate `per_dp_bs_size` / `real_bs_per_dp` / `dp_size` / `logits_indices_selector` / merged `sampling_info`
- `ModelWorkerSamplingInfo`: add `__len__` / `filter_batch`
- scheduler spec branch: write `seq_lens`/`spec_info` via `reqs_info[0]`; `device_get` `next_token_ids` before `_extract_dp_output_ids`
- `EagleDraftInput.filter_batch`: coerce list indices to `np.asarray`
- `prepare_for_extend_after_verify`: set `logits_indices`

### Regression fixes — #940 (sampler RNG)
- `model_runner`: init `_sampler_base_rng`/`_sampler_step` for draft worker too (`initialize_jit()` needs it; only the call itself stays guarded by `is_draft_worker`)

### Regression fixes — #1015 (CompilationManager)
- `eagle_worker` precompile: `generate_model_worker_batch` → `compilation_manager._make_dummy_batch` (with `dp_size`/`per_dp_bs_size`)
- `compilation_manager` spec branch: use `ModelWorkerSamplingInfo` (not `SamplingBatchInfo`) for dummy batches

### Explicit-sharding reshard at jit-exit points
- `topk_probs_from_logits`: reshard vocab-sharded logits inside jit
- `_replicate` helper; replicate logits/hidden after target verify, `capture_for_decode`, `draft_forward` step, `draft_extend_after_verify` gather
- `build_tree_kernel_efficient`: replicate inputs at entry
- `update_eagle_lists`: enable `@jit` (inputs now replicated)
- `@jax.jit(static_argnames=…)` → `@functools.partial(jax.jit, …)`
- `verify()`: redirect rejected `accept_index=-1` to per-req last slot (bs>1 prefix-KV corruption)
- `forward_draft_extend`: `device_get` `verified_id` before slicing

### `llama_eagle3` (dense EAGLE3 draft model)
- reshard `spec_info.hidden_states` to `embeds`' sharding before concat

### CI
- add `test_speculative_decoding.py` to `accuracy-test-tpu-v6e-4` suite
- test config tuned for current state (`TODO #1053 P1-2` to restore):
  - `max-running-requests` 256→64 — #1015 auto-appends max-bs to precompile buckets; `build_tree` at bs=256 OOMs on v6e-4
  - `check_cache_miss=False` + `num_examples` 512→64 — spec precompile shapes don't yet cover all runtime buckets; every new prefill-len recompiles, making 512 too slow for CI

## Verification

v6e-4 (`ct6e-standard-4t`, tp=4), fresh pod from rebuilt image:

```
Ran 1 test in 1541.989s
OK
Score: 0.766  (>0.45 threshold)
```

## Known issues (out of P1-0 scope, tracked in #1053)

- accept-len ~1.05 at bs≥4 (P1-4 known pitfall: prefix-KV bf16 drift after batched-prefill merge)
- spec precompile shape coverage incomplete → jit cache miss on each new prefill token-count (P1-2)
- `eagle_worker.add_logprob_values` is dead code (no callers) and references stale `ScheduleBatch` flat attrs; left in place for P1-2 class refactor to clean up

## Test plan

- [x] `pre-commit run` clean
- [x] `test_speculative_decoding.py` passes on v6e-4 (score 0.766)
- [ ] CI `accuracy-test-tpu-v6e-4` suite green

### CODE REVIEW

**Comment by zorrofox** [comment]
`e2e-test-4-tpu (1)` timed out after `test_server_pause_continue.py` hung for 96min. Looks flaky — #1066 (which is stacked on this exact commit) passed the same check in 31m. Could a maintainer retry the failed job?

File: python/sgl_jax/srt/speculative/eagle_worker.py [hunk 1]
```diff
@@ -470,9 +481,16 @@ def verify(self, model_worker_batch: ModelWorkerBatch, cur_allocate_lens: jax.Ar
...
         )
-        logits_output.next_token_logits = logits_output.next_token_logits[accept_index, :]
-        logits_output.hidden_states = logits_output.hidden_states[accept_index, :]
-        model_worker_batch.positions = model_worker_batch.positions[accept_index]
+        # accept_index uses -1 for rejected slots; gathering with -1 picks the
+        # global last element, so dext later writes rejected tokens' draft-KV at
+        # a foreign position inside each req's page (corrupts prefix KV for all
+        # but the last req at bs>1). Redirect -1 to each req's own last slot.
+        draft_n = self.speculative_num_draft_tokens
+        per_req_last = (np.arange(len(accept_index)) // draft_n) * draft_n + draft_n - 1
```
Comment (pengchengneo) [inline python/sgl_jax/srt/speculative/eagle_worker.py]: this per_req_last always need draft_n == spec_steps + 1 ?

Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/eagle_worker.py]: maybe better like below

```
  draft_n = self.speculative_num_draft_tokens
  accept_width = self.speculative_num_steps + 1

  req_ids = np.arange(len(accept_index)) // accept_width
  per_req_last = req_ids * draft_n + draft_n - 1
  safe_index = np.where(accept_index >= 0, accept_index, per_req_last)
```

File: python/sgl_jax/srt/speculative/eagle_worker.py [hunk 2]
```diff
@@ -778,6 +801,9 @@ def topk_probs_from_logits(
 ) -> tuple[jax.Array, jax.Array]:
     """Return top-k probabilities without materializing the full softmax tensor."""
     working_logits = jnp.moveaxis(logits, axis, -1) if axis != -1 else logits
+    sh = jax.typeof(working_logits).sharding
+    if isinstance(sh, NamedSharding):
+        working_logits = jax.sharding.reshard(working_logits, NamedSharding(sh.mesh, P()))
```
Comment (pengchengneo) [inline python/sgl_jax/srt/speculative/eagle_worker.py:812]: This reshard makes sense as a tactical dp=1 fix: `topk_probs_from_logits` needs global top-k/logsumexp over the vocab axis, while logits can arrive as `P("data", "tensor")`. 
Replicating avoids explicit-sharding errors and preserves correctness. 
Longer term, this should probably become a sharded top-k/logsumexp path instead of all-gathering full vocab logits, especially for DP/tensor-parallel scalability.

File: python/sgl_jax/srt/speculative/eagle_worker.py (hunk 1 above)
Comment (zorrofox) [reply python/sgl_jax/srt/speculative/eagle_worker.py]: Verified — `accept_index` is `(bs, spec_steps+1).flatten()` (eagle_util.py:931/993), so at topk=1 `draft_n==spec_steps+1` and the original happened to work, but it's wrong for topk>1. Adopted your version in 40dfe394.

File: python/sgl_jax/srt/speculative/eagle_worker.py (hunk 2 above)
Comment (zorrofox) [reply python/sgl_jax/srt/speculative/eagle_worker.py:812]: Agreed; added a `TODO(#1053 Phase 2)` here pointing at sharded top-k/logsumexp. Keeping the all-gather as the dp=1 baseline for now.

File: python/sgl_jax/srt/speculative/eagle_worker.py (hunk 1 above)
Comment (zorrofox) [reply python/sgl_jax/srt/speculative/eagle_worker.py]: (correction: d51d8651, not 40dfe394 — earlier push was a no-op amend)

**Comment by pengchengneo** [comment]
LGTM

**Review by pengchengneo — APPROVED** [review]
